### PR TITLE
plugin/synaptics-mst: add support for cayenne/spyder

### PR DESCRIPTION
**Pull request by ApolloLing**
Type of pull request:

- [ ] Feature

### CODE REVIEW

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 1]
```diff
@@ -771,19 +772,133 @@ fu_synaptics_mst_device_panamera_prepare_write (FuSynapticsMstDevice *self, GErr
...
+									 NULL, 4, (guint8 *)(&flash_checksum),
+									 error)) {
+			g_prefix_error (error, "Failed to get flash checksum: ");
+			return FALSE;
+		}
+		if (checksum == flash_checksum)
+			break;
+		g_debug ("attempt %u: checksum %x didn't match %x", retries_cnt, flash_checksum, checksum);
+
+
```
Comment (hughsie) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:851]: too much whitespace

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 2]
```diff
@@ -771,19 +772,133 @@ fu_synaptics_mst_device_panamera_prepare_write (FuSynapticsMstDevice *self, GErr
...
+					checksum);
+			return FALSE;
+		}
+	}
+
+	if (!fu_synaptics_mst_connection_rc_set_command (connection,
+							 UPDC_ACTIVATE_FIRMWARE,
+							 0, 0, NULL,
+							 error)) {
+		g_prefix_error (error, "can't active firmware!!! ");
```
Comment (hughsie) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:867]: maybe something like `cannot activate firmware: ` instead

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 3]
```diff
@@ -871,14 +997,21 @@ fu_synaptics_mst_device_write_firmware (FuDevice *device,
 			g_prefix_error (error, "Firmware update failed: ");
 			return FALSE;
 		}
-	} else {
-		if (!fu_synaptics_mst_device_update_tesla_leaf_firmware (self,
+		break;
+	case FU_SYNAPTICS_MST_FAMILY_CAYENNE:
+	case FU_SYNAPTICS_MST_FAMILY_SPYDER:
+	g_debug(" === Update cayenne/spyder firmware === \n");
```
Comment (hughsie) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:1003]: remove this debug line pls

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 4]
```diff
@@ -871,14 +997,21 @@ fu_synaptics_mst_device_write_firmware (FuDevice *device,
...
+		if (!fu_synaptics_mst_device_update_cayenne_firmware (self,
 									 payload_len,
 									 payload_data,
 									 error)) {
 			g_prefix_error (error, "Firmware update failed: ");
 			return FALSE;
 		}
+		break;
+	default:
+		g_prefix_error (error, "Unsupport chip family: ");
```
Comment (hughsie) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:1013]: you need to use `g_set_error()` if there's no existing error set.

**Comment by hughsie** [comment]
I think you need to run `./contrib/setup` and then use `./contrib/reformat-code.py` to fix the code style.

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 5]
```diff
@@ -940,11 +1075,26 @@ fu_synaptics_mst_device_read_board_id (FuSynapticsMstDevice *self,
...
+	case FU_SYNAPTICS_MST_FAMILY_LEAF:
+	case FU_SYNAPTICS_MST_FAMILY_PANAMERA:
+		offset = (gint)ADDR_MEMORY_CUSTOMER_ID;
+		break;
+	case FU_SYNAPTICS_MST_FAMILY_CAYENNE:
+	case FU_SYNAPTICS_MST_FAMILY_SPYDER:
+		offset = (gint)ADDR_MEMORY_CUSTOMER_ID_CAYENNE;
+		break;
+	default:
+		offset = (gint)ADDR_MEMORY_CUSTOMER_ID; //Tesla and Leaf ?
```
Comment (superm1) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:1089]: his should be an error of unknown family (ie g_set_error and return false)

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 6]
```diff
@@ -771,19 +772,133 @@ fu_synaptics_mst_device_panamera_prepare_write (FuSynapticsMstDevice *self, GErr
...
 
+static gboolean
+fu_synaptics_mst_device_update_cayenne_firmware (FuSynapticsMstDevice *self,
+						    guint32 payload_len,
+						    const guint8 *payload_data,
+						    GError **error)
+{
+	g_autoptr(FuSynapticsMstConnection) connection = NULL;
+	guint8 bytedata[4];
+	gboolean mapping_enabled = FALSE;
```
Comment (superm1) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:783]: both of these variables are not used, and CI fails from them

**Comment by superm1** [comment]
### Thanks for your commit
The fwupd project has had a style change starting with fwupd 1.6.x.  The style is summarized in https://github.com/fwupd/fwupd/blob/master/CONTRIBUTING.md and it appears your current pull request is inconsistent with the new style.  However to keep bisect useful code in the project is only changed to the new style when pull requests are opened.

**Clang Format**
Fortunately we have added a `.clang-format` and various helper scripts to help you fix your pull request.  Follow these steps to configure the environment and fix your code.
1. Run `./contrib/setup` to set up [pre-commit](https://pre-commit.com/) locally.  This will help ensure that you have all the build dependencies necessary and automatically reformat code in the future.
2. Run `./contrib/reformat-code.py master` to reformat all code that has changed in your local branch relative to master
3. Run `git commit add .; git commit --amend` or use `git rebase -i` to amend your commit(s).
 
**Pre-commit overview**
The pre-commit rules are available at https://github.com/fwupd/fwupd/blob/master/.pre-commit-config.yaml and will check all languages and text for style.  After you have run `./contrib/setup` any future commits will check the diff relative to all of these rules.

**Comment by ApolloLing** [comment]
Hi hughsie & superm1,
  Many thanks for your reviews.
  Now I have done the modifications based on your reviews and suggestions.
  I have git push the commit to https://github.com/apolloling/fwupd
  What do I need to continue the PR ?

File: plugins/synaptics-mst/fu-synaptics-mst-device.c (hunk 6 above)
Comment (ApolloLing) [reply plugins/synaptics-mst/fu-synaptics-mst-device.c]: Hi superm1,
Many thanks for your reviews.
Now I have done the modifications based on your reviews and suggestions.
I have git push the commit to https://github.com/apolloling/fwupd
What do I need to continue the PR ?

File: plugins/synaptics-mst/fu-synaptics-mst-device.c (hunk 4 above)
Comment (ApolloLing) [reply plugins/synaptics-mst/fu-synaptics-mst-device.c]: Hi hughsie,
Many thanks for your reviews.
Now I have done the modifications based on your reviews and suggestions.
I have git push the commit to https://github.com/apolloling/fwupd
What do I need to continue the PR ?

**Comment by superm1** [comment]
Can you please reformat the code using `./contrib/reformat-code.py` and then squash your commits together?

**Comment by ApolloLing** [comment]
Hi superm1,
  I reverted the previous 2 commits, then use 
./contrib/reformat-code.py
to reformat, then git push all of them.
Is it OK now ? If still not, then I think I should fork a new git for PR.

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 7]
```diff
@@ -940,12 +1090,31 @@ fu_synaptics_mst_device_read_board_id (FuSynapticsMstDevice *self,
...
+	case FU_SYNAPTICS_MST_FAMILY_CAYENNE:
+	case FU_SYNAPTICS_MST_FAMILY_SPYDER:
+		offset = (gint)ADDR_MEMORY_CUSTOMER_ID_CAYENNE;
+		break;
+	default:
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "Unsupported chip family");
+		break;
```
Comment (superm1) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:1108]: you need to return false here instead of break

**Comment by superm1** [comment]
You got the code formatted properly in that commit, just rebase and squash them together into a single commit.
https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 8]
```diff
@@ -771,20 +772,147 @@ fu_synaptics_mst_device_panamera_prepare_write (FuSynapticsMstDevice *self, GErr
...
+	case FU_SYNAPTICS_MST_FAMILY_LEAF:
+	case FU_SYNAPTICS_MST_FAMILY_PANAMERA:
+		offset = 0x2000FC;
+		break;
+	case FU_SYNAPTICS_MST_FAMILY_CAYENNE:
+	case FU_SYNAPTICS_MST_FAMILY_SPYDER:
+		offset = 0x2020021C;
+		break;
+	default:
+		offset = 0x2020021C;
```
Comment (superm1) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:904]: shouldn't this be an error?  Other areas you seem to error out on unknown family

File: plugins/synaptics-mst/fu-synaptics-mst-device.c [hunk 9]
```diff
@@ -771,20 +772,150 @@ fu_synaptics_mst_device_panamera_prepare_write (FuSynapticsMstDevice *self, GErr
...
+		}
+	}
+
+	if (!fu_synaptics_mst_connection_rc_set_command(connection,
+							UPDC_ACTIVATE_FIRMWARE,
+							0,
+							0,
+							NULL,
+							error)) {
+		g_prefix_error(error, "active firmware failed!!! ");
```
Comment (hughsie) [inline plugins/synaptics-mst/fu-synaptics-mst-device.c:878]: `activate firmware failed: ` then LGTM.

**Review by superm1 — APPROVED** [review]

**Pull request merged**